File: crypto/hpke/dispatch.go
// Package hpke implements unified dispatch for sealing and opening
// version 1 cryptographic envelopes across classical (X25519) and
// post-quantum hybrid (X-Wing) suites per BYOS-SPEC-CRYPTO-ENV-V1.
package hpke

import (
	"fmt"

	"byos.email/crypto/core"
	"byos.email/crypto/envelope"
)

// RecipientKey encapsulates the recipient's algorithm identifier and public key bytes.
type RecipientKey struct {
	Algorithm string // "HPKE-X25519-AES256GCM-v1" or "HPKE-XWing-AES256GCM-v1"
	PublicKey []byte // 32 bytes for X25519, 1216 bytes for X-Wing
}

// Seal encrypts plaintext for recipient using the algorithm specified in recipient.Algorithm.
//
// Dispatches to SealX25519 or SealXWing based on recipient.Algorithm:
//   - Unknown Algorithm returns ErrCodeUnsupportedAlgorithm (1002).
//   - Public key length mismatch returns ErrCodeInvalidKeyLength (1007).
func Seal(recipient RecipientKey, plaintext []byte, keyID string, keyEpoch uint32) (*envelope.CryptoEnvelope, error) {
	switch recipient.Algorithm {
	case envelope.AlgX25519:
		if len(recipient.PublicKey) != envelope.X25519EncSize {
			return nil, envelope.NewEnvelopeError(
				envelope.ErrCodeInvalidKeyLength,
				fmt.Sprintf("X25519 public key must be %d bytes, got %d", envelope.X25519EncSize, len(recipient.PublicKey)),
			)
		}
		return SealX25519(recipient.PublicKey, plaintext, keyID, keyEpoch)

	case envelope.AlgXWing:
		if len(recipient.PublicKey) != core.EncapsulationKeySize {
			return nil, envelope.NewEnvelopeError(
				envelope.ErrCodeInvalidKeyLength,
				fmt.Sprintf("X-Wing public key must be %d bytes, got %d", core.EncapsulationKeySize, len(recipient.PublicKey)),
			)
		}
		return SealXWing(recipient.PublicKey, plaintext, keyID, keyEpoch)

	default:
		return nil, envelope.NewEnvelopeError(
			envelope.ErrCodeUnsupportedAlgorithm,
			fmt.Sprintf("unsupported algorithm: %q", recipient.Algorithm),
		)
	}
}

// Open decrypts and verifies env using recipient's privateKey based on env.Alg.
//
//   - env is validated via envelope.Validate (rejecting malformed envelopes, unpadded base64, etc.).
//   - Unknown Algorithm returns ErrCodeUnsupportedAlgorithm (1002).
//   - Private key length mismatch returns ErrCodeInvalidKeyLength (1007).
//   - Dispatches to OpenX25519 or OpenXWing based on env.Alg.
func Open(env *envelope.CryptoEnvelope, privateKey []byte) ([]byte, error) {
	if env == nil {
		return nil, envelope.NewEnvelopeError(envelope.ErrCodeSerializationError, "envelope is nil")
	}
	if err := envelope.Validate(env); err != nil {
		return nil, err
	}

	switch env.Alg {
	case envelope.AlgX25519:
		if len(privateKey) != envelope.X25519EncSize {
			return nil, envelope.NewEnvelopeError(
				envelope.ErrCodeInvalidKeyLength,
				fmt.Sprintf("X25519 private key must be %d bytes, got %d", envelope.X25519EncSize, len(privateKey)),
			)
		}
		return OpenX25519(privateKey, env)

	case envelope.AlgXWing:
		if len(privateKey) != core.DecapsulationKeySize {
			return nil, envelope.NewEnvelopeError(
				envelope.ErrCodeInvalidKeyLength,
				fmt.Sprintf("X-Wing private key seed must be %d bytes, got %d", core.DecapsulationKeySize, len(privateKey)),
			)
		}
		return OpenXWing(privateKey, env)

	default:
		return nil, envelope.NewEnvelopeError(
			envelope.ErrCodeUnsupportedAlgorithm,
			fmt.Sprintf("unsupported algorithm: %q", env.Alg),
		)
	}
}

File: crypto-core/src/dispatch.rs
use crate::core;
use crate::envelope::{self, CryptoEnvelope, ALG_X25519, ALG_XWING, X25519_ENC_SIZE};
use crate::error::CryptoError;
use crate::hpke::{open_x25519, open_xwing, seal_x25519, seal_xwing};

/// RecipientKey encapsulates the recipient's algorithm identifier and public key bytes.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RecipientKey {
    pub algorithm: String,
    pub public_key: Vec<u8>,
}

/// Seals plaintext for recipient using the algorithm specified in recipient.algorithm.
pub fn seal(
    recipient: &RecipientKey,
    plaintext: &[u8],
    key_id: &str,
    key_epoch: u32,
) -> Result<CryptoEnvelope, CryptoError> {
    match recipient.algorithm.as_str() {
        ALG_X25519 => {
            if recipient.public_key.len() != X25519_ENC_SIZE {
                return Err(CryptoError::InvalidKeyLength);
            }
            seal_x25519(&recipient.public_key, plaintext, key_id, key_epoch)
        }
        ALG_XWING => {
            if recipient.public_key.len() != core::ENCAPSULATION_KEY_SIZE {
                return Err(CryptoError::InvalidKeyLength);
            }
            seal_xwing(&recipient.public_key, plaintext, key_id, key_epoch)
        }
        _ => Err(CryptoError::UnsupportedAlgorithm),
    }
}

/// Opens and decrypts an envelope using the recipient's private key based on env.alg.
pub fn open(
    env: &CryptoEnvelope,
    private_key: &[u8],
) -> Result<Vec<u8>, CryptoError> {
    envelope::validate(env)?;

    match env.alg.as_str() {
        ALG_X25519 => {
            if private_key.len() != X25519_ENC_SIZE {
                return Err(CryptoError::InvalidKeyLength);
            }
            open_x25519(private_key, env)
        }
        ALG_XWING => {
            if private_key.len() != core::DECAPSULATION_KEY_SIZE {
                return Err(CryptoError::InvalidKeyLength);
            }
            open_xwing(private_key, env)
        }
        _ => Err(CryptoError::UnsupportedAlgorithm),
    }
}
